pdns/packetcache.hh

```diff
@@ -23,80 +23,137 @@
 #include "ednsoptions.hh"
 #include "misc.hh"
 #include "iputils.hh"
+#include "views.hh"
 
 class PacketCache : public boost::noncopyable
 {
 public:
-  static uint32_t canHashPacket(const std::string& packet, uint16_t* ecsBegin, uint16_t* ecsEnd)
+
+  /* hash the packet from the provided position, which should point right after tje qname. This skips:
+     - the query ID ;
+     - EDNS Cookie options, if any ;
+     - EDNS Client Subnet options, if any and skipECS is true.
+  */
+  static uint32_t hashAfterQname(const string_view& packet, uint32_t currentHash, size_t pos, bool skipECS)
   {
-    uint32_t ret = 0;
-    ret = burtle(reinterpret_cast<const unsigned char*>(packet.c_str()) + 2, sizeof(dnsheader) - 2, ret); // rest of dnsheader, skip id
-    size_t packetSize = packet.size();
-    size_t pos = sizeof(dnsheader);
-    const char* end = packet.c_str() + packetSize;
-    const char* p = packet.c_str() + pos;
-
-    for(; p < end && *p; ++p, ++pos) { // XXX if you embed a 0 in your qname we'll stop lowercasing there
-      const unsigned char l = dns_tolower(*p); // label lengths can safely be lower cased
-      ret=burtle(&l, 1, ret);
-    }                           // XXX the embedded 0 in the qname will break the subnet stripping
-
-    const struct dnsheader* dh = reinterpret_cast<const struct dnsheader*>(packet.c_str());
-    const char* skipBegin = p;
-    const char* skipEnd = p;
-    if (ecsBegin != nullptr && ecsEnd != nullptr) {
-      *ecsBegin = 0;
-      *ecsEnd = 0;
-    }
-    /* we need at least 1 (final empty label) + 2 (QTYPE) + 2 (QCLASS)
+    const size_t packetSize = packet.size();
+    assert(packetSize >= sizeof(dnsheader));
+
+    /* we need at least 2 (QTYPE) + 2 (QCLASS)
+
        + OPT root label (1), type (2), class (2) and ttl (4)
        + the OPT RR rdlen (2)
-       = 16
+       = 15
     */
-    if(ntohs(dh->arcount)==1 && (pos+16) < packetSize) {
-      char* optionBegin = nullptr;
-      size_t optionLen = 0;
-      /* skip the final empty label (1), the qtype (2), qclass (2) */
-      /* root label (1), type (2), class (2) and ttl (4) */
-      int res = getEDNSOption(const_cast<char*>(reinterpret_cast<const char*>(p)) + 14, end - (p + 14), EDNSOptionCode::ECS, &optionBegin, &optionLen);
-      if (res == 0) {
-        skipBegin = optionBegin;
-        skipEnd = optionBegin + optionLen;
-        if (ecsBegin != nullptr && ecsEnd != nullptr) {
-          *ecsBegin = optionBegin - packet.c_str();
-          *ecsEnd = *ecsBegin + optionLen;
-        }
+    const struct dnsheader* dh = reinterpret_cast<const struct dnsheader*>(packet.data());
+    if (ntohs(dh->qdcount) != 1 || ntohs(dh->ancount) != 0 || ntohs(dh->nscount) != 0 || ntohs(dh->arcount) != 1 || (pos + 15) >= packetSize) {
```

Comment:
You could choose to `htons()` the right hand side of the compares here, since they all test for inequality.

Reply:
I'd really prefer to keep it that way as I believe it's more easy to read than the alternative you suggest. Even if the compiler were somehow not smart enough to optimize it, I'm sure we can live with it.